contracts/test/PlonkVerifier.t.sol

```diff
@@ -0,0 +1,378 @@
+    function test_correctConstants() external {
+        string[] memory cmds = new string[](3);
+        cmds[0] = "diff-test";
+        cmds[1] = "plonk-constants";
+
+        bytes memory result = vm.ffi(cmds);
+        (
+            uint256 k1,
+            uint256 k2,
+            uint256 k3,
+            uint256 k4,
+            uint256 betaHX0,
+            uint256 betaHX1,
+            uint256 betaHY0,
+            uint256 betaHY1
+        ) = abi.decode(
+            result, (uint256, uint256, uint256, uint256, uint256, uint256, uint256, uint256)
+        );
+
+        assertEq(k1, V.COSET_K1);
+        assertEq(k2, V.COSET_K2);
+        assertEq(k3, V.COSET_K3);
+        assertEq(k4, V.COSET_K4);
+        assertEq(betaHX0, V.BETA_H_X0);
+        assertEq(betaHX1, V.BETA_H_X1);
+        assertEq(betaHY0, V.BETA_H_Y0);
+        assertEq(betaHY1, V.BETA_H_Y1);
+    }
+}
+
+contract PlonkVerifier_batchVerify_Test is Test {
+    /// forge-config: default.fuzz.runs = 30
+    /// @dev Test if some of the user inputs are invalid
+    function testFuzz_revertWhenInvalidArgs(
+        IPlonkVerifier.VerifyingKey[] memory verifyingKeys,
+        uint256[][] memory publicInputs,
+        IPlonkVerifier.PlonkProof[] memory proofs,
+        bytes[] memory extraTranscriptInitMsgs
+    ) external {
+        vm.assume(
+            verifyingKeys.length != proofs.length || publicInputs.length != proofs.length
+                || extraTranscriptInitMsgs.length != proofs.length || proofs.length == 0
+        );
+        vm.expectRevert(V.InvalidPlonkArgs.selector);
+        V.batchVerify(verifyingKeys, publicInputs, proofs, extraTranscriptInitMsgs);
+    }
+
+    /// @dev Test happy and unhappy path of `batchVerify`.
+    function test_batchVerify_succeeds() external {
+        for (uint32 i = 1; i < 6; i++) {
+            string[] memory cmds = new string[](3);
+            cmds[0] = "diff-test";
+            cmds[1] = "plonk-batch-verify";
+            cmds[2] = vm.toString(i);
+
+            bytes memory result = vm.ffi(cmds);
+            (
+                IPlonkVerifier.VerifyingKey[] memory verifyingKeys,
+                uint256[][] memory publicInputs,
+                IPlonkVerifier.PlonkProof[] memory proofs,
+                bytes[] memory extraTranscriptInitMsgs
+            ) = abi.decode(
+                result,
+                (IPlonkVerifier.VerifyingKey[], uint256[][], IPlonkVerifier.PlonkProof[], bytes[])
+            );
+
+            // happy path
+            assert(V.batchVerify(verifyingKeys, publicInputs, proofs, extraTranscriptInitMsgs));
+
+            // unhappy path
+            // wrong vk
+            IPlonkVerifier.VerifyingKey[] memory badVks = verifyingKeys;
+            badVks[0].q1 = BN254.negate(verifyingKeys[0].q1);
+            badVks[0].qEcc = BN254.negate(verifyingKeys[0].qEcc);
+            assert(!V.batchVerify(badVks, publicInputs, proofs, extraTranscriptInitMsgs));
```

Comment:
I read "[data location and assignment behavior](https://docs.soliditylang.org/en/latest/types.html#data-location-and-assignment-behavior)" today and realize that this test logic is wrong.

cuz the new `badVks` mutation will also affect the underlying `verifyingKeys` since they are both marked as `memory`, thus the following bad tests are not isolating the bad input arguments.

I will work on this next working day.

Reply:
fixed in [bdf18ac](https://github.com/EspressoSystems/espresso-sequencer/pull/722/commits/bdf18acddacaf8702c3c403bfb9346f1cd1c9056)